File: python/leetcode/124.py
class Solution:
    def maxsum(self, root):
        if not root: return 0
        left = max(0, self.maxsum(root.left))
        right = max(0, self.maxsum(root.right))
        Solution.max = max(Solution.max, left + right + root.val)
        return max(left, right) + root.val

    def maxPathSum(self, root):
        Solution.max = -float("Infinity")
        self.maxsum(root)
        return Solution.max

File: cpp/leetcode/124.cpp
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int res = INT_MIN;
        dfs(root, res);
        return res;
    }
    
    int dfs(TreeNode *root, int &res){
        if(!root) return 0;
        int left = max(0, dfs(root->left, res));
        int right = max(0, dfs(root->right, res));
        res = max(res, left + right + root->val);
        return max(left, right) + root->val;
    }
};